### ISSUE TITLE
[Configuration] Should we not distinguish between singleton/factory?

### ISSUE THREAD

**Issue by rinswind** [pkriens/cdi#24]
- **For**
    - Flexible add hoc configuration like in DS
        - Singleton: consume one regardless of how many are available
        - Factory: consume as many as available
    - Less error prone: People forget to make a factory configuration when they need one instance.
- **Against**
    - Confusing given that the Config Admin spec is so split into singleton vs factory

**Comment by rotty3000** [comment]
Please explain more. I'm not sure what this is talking about.

**Comment by rinswind** [comment]
Don't register a `ManagedService` for singleton configurations.
Don't register a `ManagedServiceFactory` for factory configurations.
Track configurations with a `ConfigurationListener` disregarding their cardinality (one/many) as seen in ConfigurationAdmin.
For configurations declared with `@PID` consume only one object even if it is a factory configuration.
For configurations declared with `@FactoryPID` consume many objects if it is a factory and one if it is a singleton. I.e. consume as many as we can find.
Isn't this how DS flip/flops?

**Comment by rotty3000** [comment]
> Don't register a ManagedService for singleton configurations.
> Don't register a ManagedServiceFactory for factory configurations.
> Track configurations with a ConfigurationListener disregarding their cardinality (one/many) as seen in ConfigurationAdmin.
> For configurations declared with @PID consume only one object even if it is a factory configuration.
I disagree with this! I think if the configuration is NOT a singleton, it should be ignored, otherwise you end up with guessing which configuraton instance is the one being used.
> For configurations declared with @FactoryPID consume many objects if it is a factory and one if it is a singleton.
I also disagree with this! Only consumer configurations if they are instances of a factory.

One of the major points in this regard is that DS has a very odd behaviour w.r.t. factory configurations and CDI was meaning to eliminate that odd behaviour.

**Comment by rinswind** [comment]
I don't like the DS way and always wondered why it was introduced at all. 
I opened the issue because it was not discussed formally - we just decided this is a bad idea.
If anyone can tell me what are the benefits for DS we can discuss if we need them in CDI.

Actually the in DS there is no "structural" need to separate factory from singleton like we have in CDI. So DS can create as many components as there are configurations - can't make it create just one. So I suppose I have to close this one because of the ambiguity with picking a single configuration out of a set of factory objects you pointed out.
